File: calculator_server.py
#!/usr/bin/env python3
"""
Simple MCP Calculator Server

This server exposes mathematical operations as MCP tools.
"""
from mcp.server.fastmcp import FastMCP, Context
import logging
logging.basicConfig(level=logging.DEBUG)

# Create our MCP server
mcp = FastMCP("Calculator", debug=True)

@mcp.tool()
def add(a: float, b: float) -> float:
    """Add two numbers.
    
    Args:
        a: First number
        b: Second number
        
    Returns:
        The sum of a and b
    """
    return a + b

@mcp.tool()
def subtract(a: float, b: float) -> float:
    """Subtract b from a.
    
    Args:
        a: First number
        b: Second number
        
    Returns:
        The result of a - b
    """
    return a - b

@mcp.tool()
def multiply(a: float, b: float) -> float:
    """Multiply two numbers.
    
    Args:
        a: First number
        b: Second number
        
    Returns:
        The product of a and b
    """
    return a * b

@mcp.tool()
def divide(a: float, b: float) -> float:
    """Divide a by b.
    
    Args:
        a: Numerator
        b: Denominator (must not be zero)
        
    Returns:
        The result of a / b
        
    Raises:
        ValueError: If b is zero
    """
    if b == 0:
        raise ValueError("Cannot divide by zero")
    return a / b

@mcp.tool()
def power(base: float, exponent: float) -> float:
    """Calculate base raised to the power of exponent.
    
    Args:
        base: The base number
        exponent: The exponent
        
    Returns:
        base ^ exponent
    """
    return base ** exponent

@mcp.tool()
def calculate(expression: str) -> float:
    """Evaluate a mathematical expression.
    
    Args:
        expression: A mathematical expression string
        
    Returns:
        The result of evaluating the expression
        
    Note:
        Uses Python's eval function with limited scope for basic calculations
    """
    # For security, only allow basic math operations with a limited scope
    allowed_names = {"abs": abs, "max": max, "min": min, "pow": pow, "round": round}
    
    try:
        # Evaluate the expression with limited scope
        result = eval(expression, {"__builtins__": {}}, allowed_names)
        return float(result)
    except Exception as e:
        raise ValueError(f"Invalid expression: {str(e)}")

@mcp.resource("calculator://help")
def get_calculator_help() -> str:
    """Provide help information about the calculator.
    
    Returns:
        Help text describing the calculator functions
    """
    return """
    Calculator MCP Server
    
    This server provides basic mathematical operations:
    
    - add(a, b) - Add two numbers
    - subtract(a, b) - Subtract b from a
    - multiply(a, b) - Multiply two numbers
    - divide(a, b) - Divide a by b (b cannot be zero)
    - power(base, exponent) - Calculate base raised to the power of exponent
    - calculate(expression) - Evaluate a mathematical expression
    
    Examples:
      add(5, 3) -> 8
      subtract(10, 4) -> 6
      multiply(2.5, 4) -> 10
      divide(10, 2) -> 5
      power(2, 3) -> 8
      calculate("10 + 5 * 2") -> 20
    """

@mcp.prompt()
def calculation_prompt(expression: str) -> str:
    """Create a prompt for calculating a mathematical expression.
    
    Args:
        expression: The mathematical expression to evaluate
        
    Returns:
        A prompt for calculating the expression
    """
    return f"""Please evaluate this mathematical expression: {expression}
    
You can use the calculator tools if needed.
"""

if __name__ == "__main__":
    # Run the server
    mcp.run()

File: mcp-calculator/calculator_server.js
#!/usr/bin/env node
/**
 * Simple MCP Calculator Server
 * 
 * This server exposes mathematical operations as MCP tools.
 */
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import { z } from "zod";

// Create server instance
const server = new McpServer({
  name: "Calculator",
  version: "1.0.0",
});

// Add tool
server.tool(
  "add",
  "Add two numbers",
  {
    a: z.number().describe("First number"),
    b: z.number().describe("Second number"),
  },
  async ({ a, b }) => {
    return {
      content: [
        {
          type: "text",
          text: "42",//String(a + b),
        },
      ],
    };
  }
);

// Subtract tool
server.tool(
  "subtract",
  "Subtract b from a",
  {
    a: z.number().describe("First number"),
    b: z.number().describe("Second number"),
  },
  async ({ a, b }) => {
    return {
      content: [
        {
          type: "text",
          text: String(a - b),
        },
      ],
    };
  }
);

// Multiply tool
server.tool(
  "multiply",
  "Multiply two numbers",
  {
    a: z.number().describe("First number"),
    b: z.number().describe("Second number"),
  },
  async ({ a, b }) => {
    return {
      content: [
        {
          type: "text",
          text: String(a * b),
        },
      ],
    };
  }
);

// Divide tool
server.tool(
  "divide",
  "Divide a by b",
  {
    a: z.number().describe("Numerator"),
    b: z.number().describe("Denominator (must not be zero)"),
  },
  async ({ a, b }) => {
    if (b === 0) {
      return {
        isError: true,
        content: [
          {
            type: "text",
            text: "Error: Cannot divide by zero",
          },
        ],
      };
    }
    
    return {
      content: [
        {
          type: "text",
          text: String(a / b),
        },
      ],
    };
  }
);

// Power tool
server.tool(
  "power",
  "Calculate base raised to the power of exponent",
  {
    base: z.number().describe("The base number"),
    exponent: z.number().describe("The exponent"),
  },
  async ({ base, exponent }) => {
    return {
      content: [
        {
          type: "text",
          text: String(Math.pow(base, exponent)),
        },
      ],
    };
  }
);

// Calculate expression tool
server.tool(
  "calculate",
  "Evaluate a mathematical expression",
  {
    expression: z.string().describe("A mathematical expression string"),
  },
  async ({ expression }) => {
    try {
      // Note: In a production environment, you would want to use
      // a safer expression evaluator than eval
      const result = eval(expression);
      
      return {
        content: [
          {
            type: "text",
            text: String(result),
          },
        ],
      };
    } catch (error) {
      return {
        isError: true,
        content: [
          {
            type: "text",
            text: `Error: Invalid expression - ${error.message}`,
          },
        ],
      };
    }
  }
);

// Add a resource
server.resource({
  uri: "calculator://help",
  name: "Calculator Help",
  description: "Help information about the calculator",
  async read() {
    return {
      contents: [
        {
          uri: "calculator://help",
          mimeType: "text/plain",
          text: `
Calculator MCP Server

This server provides basic mathematical operations:

- add(a, b) - Add two numbers
- subtract(a, b) - Subtract b from a
- multiply(a, b) - Multiply two numbers
- divide(a, b) - Divide a by b (b cannot be zero)
- power(base, exponent) - Calculate base raised to the power of exponent
- calculate(expression) - Evaluate a mathematical expression

Examples:
  add(5, 3) -> 8
  subtract(10, 4) -> 6
  multiply(2.5, 4) -> 10
  divide(10, 2) -> 5
  power(2, 3) -> 8
  calculate("10 + 5 * 2") -> 20
          `,
        },
      ],
    };
  },
});

// Add a prompt
server.prompt({
  name: "calculation",
  description: "Create a prompt for calculating a mathematical expression",
  arguments: [
    {
      name: "expression",
      description: "The mathematical expression to evaluate",
      required: true,
    },
  ],
  async get(args) {
    const expression = args.arguments?.expression || "";
    
    return {
      messages: [
        {
          role: "user",
          content: {
            type: "text",
            text: `Please evaluate this mathematical expression: ${expression}\n\nYou can use the calculator tools if needed.`,
          },
        },
      ],
    };
  },
});

// Start server
async function main() {
  const transport = new StdioServerTransport();
  await server.connect(transport);
  console.error("Calculator MCP Server running on stdio");
}

main().catch((error) => {
  console.error("Fatal error:", error);
  process.exit(1);
});